### PR TITLE
feat: add coreset and coresubset data structures

### LINKED ISSUE #518: Add a class for coresubsets
### What's the new feature?

Add a `CoreSubset` class like that shown below:

```python
# Current implementation
class KernelHearding(coreax.reduction.Coreset):
    self.coreset_indices = ... # some value that is set after a call to `fit_to_size`

# Proposed implementation
class CoreSubset(coreax.reduction.Coreset):
      """Docstring explaining that a coresubset is a special type of coreset"""
      self.coreset_indices = None  # Remove from Coreset and init here as None

class KernelHearding(CoreSubset):
      self.coreset_indices = ... # some value that is overwritten from `None` after a call to `fit_to_size`
```

### What value does this add?

At the moment, a child of `Coreset` is considered as a coresubset if the `Coreset.corset_indices is not None`. This kind of implicit "typing" can be confusing for new users, and makes it impossible to determine the "type" of coreset algorithm being used at runtime (prior to fitting, all corsets have `Corset.corset_indices = None`)

*Documentation example:* if the `KernelHerding` class were not in the `coresubset.py` file, which may not be the case for future coresubset algorithms, the only way to determine that it is a coresubset (not a coreset) is to inspect the code in its `fit_to_size` method. This assumes the user is already aware that setting a `coreset_indices != None` indicates a coresubset algorithm. If `KernelHerding` was a child of a `CoreSubset` class, it is immediately obvious that it generates a coresubset, and that the important documentation regarding the difference between a `Coreset` and a `Coresubset` can be found by going to the definition of `CoreSubset`.

*Runtime example:* imagine a user wishes to construct a searchable set of `Coreset` classes/algorithms. Currently, there is no option to filter such a set to only `CoreSubsets`. However, if the coresubset methods all inherited from a `CoreSubset` class, it would be as simple as `filter(lambda x: isinstance(x, CoreSubset), searchable_set)`. You could even do something like [this](https://stackoverflow.com/questions/3862310/how-to-find-all-the-subclasses-of-a-class-given-its-name) if you wanted to programmatically identify all `CoreSubsets` implemented in coreax.

### Is there an alternative you've considered?

Continue with the current approach.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tc85324**
### PR Type
Closes #518
- Feature
- Tests

### Description
As part of #581, it is necessary to decouple the coreset data-structures and solvers/reduction-strategies. The new `coreax.coreset.Coreset` and `coreax.coreset.Coresubset` data-structures introduced here facilitate this decoupling, such that the conceptual notion of a coreset better reflects our implementation. I.E. a coreset is a data-structure and a solver is a solver, the current implementation treats a coreset as both of these things.

Note: these new data-structures are currently unused. They will be used in an upcoming PR to fully address #581.

A new set of unit tests have been created for these new data-structures in `tests/unit/test_coreset.py`.

Additionally updates the new `coreax.data.Data` with `at_least{1,2}d` equinox converters and a `__len__` magic method. This facilitates the coreset `__len__` magic method, and will allow many `atleast_{1,2}d` calls to be removed from the codebase, at a later date.

Tests for `coreax.data.Data` have been refactored to reduce code duplication resulting from a new test for the new `__len__` method.

### How Has This Been Tested?
All tests pass

### Does this PR introduce a breaking change?
No

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: coreax/coreset.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    return Data(data)
+
+
+class Coreset(eqx.Module):
+    r"""
+    Data structure for representing a coreset.
+
+    TLDR: a coreset is a reduced set of :math:`\hat{n}` (potentially weighted) data
+    points that, in some sense, best represent the "important" properties of a larger
+    set of :math:`n > \hat{n}` (potentially weighted) data points.
```
Comment (db091756) [inline coreax/coreset.py:24]: Could stick in a >>> here to signify that the coreset is supposed to be much smaller than n

**Review by db091756 — CHANGES_REQUESTED** [review]
All looks good to me, though I think we need a SupervisedCoreset and SupervisedCoresubset class to handle supervised coresets with the SupervisedData class

File: coreax/coreset.py (hunk 1 above)
Comment (tc85324) [reply coreax/coreset.py:24]: It depends if you would give a set where `n` is with `>` but not `>>` `\hat{n}` a different name? If not, then I think the current notation is more accuracte.

**Comment by tc85324** [comment]
If we want the ability to distinguish between supervised and unsupervised coresets, then we can make the `pre_coreset_data` generic on `_Data = TypeVar("_Data", bound=Data)`. Then you could type `Coreset[Data]` or `Coreset[SupervisedData]`. The materialized coreset is independent of the supervision, so we only need it for typing `pre_coreset_data`.

File: coreax/coreset.py (hunk 1 above)
Comment (db091756) [reply coreax/coreset.py:24]: I guess philosophically if you've got n=1,000,000 and hat(n)=999,999 then its not much of a coreset, but i take your point and happy to leave as is!

File: coreax/coreset.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        coresubset nodes should be accessed via :meth:`Coresubset.coreset`.
+    :param pre_coreset_data: The dataset :math:`X` used to construct the coreset.
+    """
+
+    # Incompatibility between Pylint and eqx.field. Pyright handles this correctly.
+    # pylint: disable=no-member
+    @property
+    def coreset(self) -> Data:
+        """Materialise the coresubset from the indices and original data."""
+        coreset_nodes = self.pre_coreset_data.data[self.unweighted_indices]
```
Comment (db091756) [inline coreax/coreset.py:110]: Maybe I am misunderstanding, but if this is a supervised Coresubset then we won't be materialising the supervision here right?

Comment (tc85324) [reply coreax/coreset.py:110]: This was by design. My logic was that the coreset should only be the weighted nodes (this is consistent with the docstrings). If you require the supervision associated with each coreset node/index, then you can get this with `Coreset.pre_coreset_data.supervision[self.unweighted_indices]`, providing the `pre_coreset_data` is `SupervisedData`.

Comment (db091756) [reply coreax/coreset.py:110]: I see, makes sense!

**Review by db091756 — APPROVED** [review]

**Pull request merged**